Segment list design [MAILPOET-5395]

websupporter opened the pull request:
## Description
This fixes [MAILPOET-5395]. Some elements are not completely like in the design. E.g. you can't sort by contacts or subscribers and the magnifier is missing in the search box. That is to be expected as for now.

## Code review notes

_N/A_

## QA notes

_N/A_

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-5395]

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-5395]: https://mailpoet.atlassian.net/browse/MAILPOET-5395?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[MAILPOET-5395]: https://mailpoet.atlassian.net/browse/MAILPOET-5395?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

rodrigoprimo committed:
Use WP i18n functions for the segment listing page

This commit improves the segment listing page by using WP i18n functions
instead of MP i18n functions. One string was deleted as it is not part
of the new design that will be implemented in subsequent commits and it
was not following i18n best practices.

[MAILPOET-5395]

rodrigoprimo committed:
Add option to <TopBar> to hide screen options

Doing this to avoid having to call <TopBarWithBeamer /> and
<HideScreenOptions /> separately. Will be used in the new design of the
segments listing page.

[MAILPOET-5395]

rodrigoprimo committed:
Implement new design of the header of the segment listing page

[MAILPOET-5395]

rodrigoprimo committed:
Implement temporary way to get dynamic segments

This will be used to populate the new dynamic segments table in the
listing page. It will need to be updated in the future as currently all
the parameters are hardcoded and I'm not sure if I'm doing things the
React way here.

[MAILPOET-5395]

rodrigoprimo committed:
Add basic functionality of the new dynamic segments listing table

[MAILPOET-5395]

websupporter committed:
Enable sorting and pagination

[MAILPOET-5395]

websupporter committed:
Seperate listing tables and tabs

[MAILPOET-5395]

websupporter committed:
Set offset to 0 when switching tab

[MAILPOET-5395]

websupporter committed:
Update query always when no query is currently stored

[MAILPOET-5395]

websupporter committed:
Read state from URL and write state to URL

[MAILPOET-5395]

websupporter committed:
Improve naming of tabs and content elements

[MAILPOET-5395]

websupporter committed:
Add search bar

[MAILPOET-5395]

websupporter committed:
Use count to determine the total found rows

[MAILPOET-5395]

websupporter committed:
Add store to select and unselect segments

[MAILPOET-5395]

websupporter committed:
Add select all functionality

[MAILPOET-5395]

websupporter committed:
Introduce Bulk Actions

[MAILPOET-5395]

websupporter committed:
Add notices and remove useless wrapper

[MAILPOET-5395]

websupporter committed:
Separate bulk action and action modal

[MAILPOET-5395]

websupporter committed:
Add row actions

[MAILPOET-5395]

websupporter committed:
Add duplicate action

[MAILPOET-5395]

websupporter committed:
Ensure integer query values are in the correct format

[MAILPOET-5395]

websupporter committed:
Remove old listing

[MAILPOET-5395]

websupporter committed:
Bring default limit to 25

[MAILPOET-5395]

websupporter committed:
Use ConfirmDialog instead of Modal

[MAILPOET-5395]

websupporter committed:
Add value to search field

[MAILPOET-5395]

websupporter committed:
Use wordpress/components for search and select

[MAILPOET-5395]

websupporter committed:
Styling of Dynamic Segment Listing

[MAILPOET-5395]

websupporter committed:
Apply prettier

johnolek commented:
@websupporter It looks like there are some pipeline failures that will need to be addressed. Do you have time to look at this again or would you like me to take it over?

websupporter committed:
Fix acceptance tests

[MAILPOET-5395]

Commit:
Refactor updateDynamicQueryFromLocation

MAILPOET-5395

Commit:
Fix plugin missing state

MAILPOET-5395

Commit:
Reload dynamic segments after save

MAILPOET-5395

Commit:
Refactor getTabFromLocation to use regex

MAILPOET-5395

Commit:
Add automation id for select all checkbox

MAILPOET-5395

Commit:
Fix ManageSegmentsCest selector issues

MAILPOET-5395

Commit:
Fix ManageWooCommerceSegmentsCest selector issues

MAILPOET-5395

Commit:
Fix other selector issues in segment tests

MAILPOET-5395

Commit:
Fix link casing and remove no longer needed check

MAILPOET-5395

Commit:
Display proper notices for segment actions

This updates some of the segment listing messages to include the name of
 the segment to avoid confusion, and allows for the possibility that
 bulk actions will work for some segments but not all, in which case we
 need to show both success and error messages.

MAILPOET-5395

johnolek commented:
@websupporter Do you think you would have time to take a look at this? I think I have it working and all the tests passing finally. I ask because you were working on it before and are already familiar with it, but if you don't have time that's totally ok and I can reassign it.

websupporter commented:
Hi @johnolek, I would like other engineers to get familiar with the context. I reassigned the PR therefore.

JanJakes commented on mailpoet/assets/css/src/mailpoet-dynamic-segments.scss:
```diff
@@ -0,0 +1,35 @@
+@import 'settings/colors';
+
+#dynamic_segments_container {
+  padding: 0 12px;
```
This seems to cause a padding that is inconsistent with other pages. It may have been designed like that, but I don't think we should do it. Look at other pages vs. this one, especially automations vs. segments. If we were to make larger paddings, then it should be done on all pages.

JanJakes commented on mailpoet/assets/css/src/mailpoet-dynamic-segments.scss:
```diff
@@ -0,0 +1,35 @@
+}
+
+.mailpoet-segment-listing-heading {
+  padding: 12px 0 32px;
```
Same here, I don't think we should add inconsistent spacing here. As for headings, I also think we should use the existing [PageHeader](https://github.com/mailpoet/mailpoet/tree/trunk/mailpoet/assets/js/src/common/page-header) component rather than introducing a new one.

JanJakes commented on mailpoet/assets/js/src/segments/dynamic/list/listing-header.tsx:
```diff
@@ -0,0 +1,42 @@
+import { SubscribersCacheMessage } from 'common/subscribers-cache-message';
+import { MssAccessNotices } from 'notices/mss-access-notices';
+
+export function ListingHeader(): JSX.Element {
```
Could we use the existing [PageHeader](https://github.com/mailpoet/mailpoet/tree/trunk/mailpoet/assets/js/src/common/page-header) component?

JanJakes commented on mailpoet/assets/js/src/segments/static/heading.tsx:
```diff
@@ -5,32 +5,19 @@ import { plusIcon } from 'common/button/icon/plus';
-import * as ROUTES from 'segments/routes';
 
-function ListHeading({ segmentType }): JSX.Element {
+function ListHeading(): JSX.Element {
```
Hmm, now I'm a bit confused what is `ListingHeader` vs. `ListHeading`, so I'm not sure if the [PageHeader](https://github.com/mailpoet/mailpoet/tree/trunk/mailpoet/assets/js/src/common/page-header) argument applies here...?

JanJakes commented on mailpoet/assets/js/src/segments/dynamic/store/actions.ts:
```diff
@@ -222,3 +223,34 @@ export function setPreviousPage(data: string): SetPreviousPageActionType {
+    })
+    .fail((response) => {
+      if (response.errors.length > 0) {
+        // TODO: handle errors
```
There is a `TODO` here, but I also wonder if `done`/`fail` is needed, or if with `await` we can just look at the returned values...

JanJakes commented on mailpoet/assets/css/src/components-automation-listing/_listing.scss:
```diff
@@ -29,45 +31,6 @@
-  background-color: #fff;
-  border: 1px solid #dcdcde;
-  border-radius: 2px;
-  margin-top: 16px;
```
I think this commit seems to cause a missing spacing on the automation listing page:

<img width="803" alt="Screenshot 2024-02-21 at 09 18 13" src="https://github.com/mailpoet/mailpoet/assets/141436/885d7588-d4ec-4740-84f6-d78818b15077">

JanJakes commented on mailpoet/assets/css/src/components-automation-listing/_listing.scss:
```diff
@@ -29,45 +31,6 @@
 
-.mailpoet-filter-tab-panel {
-  background-color: #fff;
-  border: 1px solid #dcdcde;
```
The listing in automations used to have an outer border that now disappeared. This may have been updated in Figma, but we should probably double-check if that's desired. Borderless tables or cards seem rather unusual in the WP/Woo admin design.

JanJakes commented on mailpoet/assets/js/src/segments/dynamic/list/listing-table.tsx:
```diff
@@ -0,0 +1,157 @@
+
+  useEffect(() => {
+    loadDynamicSegments().catch(() => {
+      // TODO: check if this is the right away to use loadDynamicSegments and handle errors if it is
```
There is a `TODO` here, and a small thing to note is we're generally preferring async/await over then/catch.

JanJakes commented on mailpoet/assets/js/src/segments/dynamic/list/listing-table.tsx:
```diff
@@ -0,0 +1,157 @@
+          totalRows={dynamicSegments.length}
+          query={Object.fromEntries(pageSearch)}
+          showMenu={false}
+        />
```
Do you think we could also do [this](https://github.com/mailpoet/mailpoet/pull/5404/commits/b1586403cd60372b1389fef4de6c5decdec3f94e) to have the placeholder use the real number of segments?

JanJakes commented on mailpoet/assets/js/src/segments/dynamic/list/listing-table.tsx:
```diff
@@ -1,13 +1,74 @@
+    sort_by: 'updated_at',
+    sort_order: 'desc',
+    group: 'all',
+  };
```
The default query seems to be defined exactly the same way a few lines above. Maybe we could extract & reuse it.

JanJakes replied:
Update: I think this was done later.

JanJakes commented on mailpoet/assets/js/src/segments/dynamic/list/listing-table.tsx:
```diff
@@ -1,13 +1,74 @@
+      }
+    }
+  }
+  updateDynamicQuery(query);
```
This code looks a bit complex, and I'm wondering what's happening here in general. We seem to have:
1. `updateDynamicQuery` that syncs the query to the store,
2. `updateDynamicQueryFromLocation` that does "the same", but from the URL.

It seems to me that the state is unnecessarily duplicated, and that it may be hard to sync those two (store and URL). Why don't we store the dynamic query state in URL only? It's a common pattern to use URL as the only state in those cases.

The sort of unusually looking URL format comes from the Woo component? In any case, as for the parsing, I think it would be better to parse the full path like:

```ts
const values = str.split('/')
  .map((part) => part.replace(/]$/, '').split('['))
  .reduce((map, [k, v]) => ({ ...map, [k]: v }), {});
```

To get an object like:

```ts
{
  group: "all",
  limit: "25",
  other: undefined,
  search: "asadf",
  sort_by: "updated_at",
  sort_order: "desc"
}
```

And then use that.

JanJakes commented on mailpoet/assets/js/src/segments/dynamic/store/actions.ts:
```diff
@@ -224,33 +226,58 @@ export function setPreviousPage(data: string): SetPreviousPageActionType {
       }
+    })
+    .catch(() => {
+      // @ToDo: show notice
```
A `TODO` here

JanJakes commented on mailpoet/assets/js/src/segments/dynamic/list/listing-table.tsx:
```diff
@@ -108,48 +168,108 @@ export function ListingTable(): JSX.Element {
+              updateDynamicQuery(newParams);
+            }
+          }}
+          query={tableQueryParams}
```
One thing I noticed when typing into the search field is that there is no debounce and/or request aborting. Apart from firing many requests at the same time, this can probably lead to out-of-order responses displaying wrong data.

JanJakes commented on mailpoet/assets/js/src/segments/dynamic/list/get-row.tsx:
```diff
@@ -2,13 +2,31 @@ import { Link } from 'react-router-dom';
+      return;
+    }
+    dispatch(storeName).selectDynamicSection(dynamicSegment);
+  };
```
I'm not fully sure if selection belongs to the store, but let's see later commits...

JanJakes replied:
Update: I found the confirmation modal using this, I guess it makes sense.

JanJakes commented on mailpoet/assets/js/src/segments/dynamic/list/bulk-actions.tsx:
```diff
@@ -0,0 +1,202 @@
+    },
+  }).then(() => {
+    void dispatch(storeName).loadDynamicSegments();
+  });
```
We should probably prefer `async`/`await`. Here it would have the side effect of revealing a potential request error here and not somewhere else. And yeah, errors are probably not handled here, but maybe they don't need to be, not sure.

JanJakes commented on mailpoet/assets/js/src/segments/dynamic/list/bulk-actions.tsx:
```diff
@@ -0,0 +1,202 @@
+      </Button>
+    </Modal>
+  );
+}
```
In some places, we're using `import { __experimentalConfirmDialog as ConfirmDialog } from '@wordpress/components';`, but I guess no big deal to keep it as it is now.

JanJakes replied:
Update: I think this was done later.

JanJakes commented on mailpoet/assets/js/src/segments/dynamic/list/bulk-actions.tsx:
```diff
@@ -0,0 +1,202 @@
+      message = sprintf(
+        _n(
+          'Are you sure you want to trash the selected segment %s?',
+          'Are you sure you want to trash the selected segments %s and %s?',
```
A small improvement: `%s and %s` => `%s, and %s` (+ for all other messages below)

JanJakes commented on mailpoet/assets/js/src/segments/dynamic/list/bulk-actions.tsx:
```diff
@@ -0,0 +1,202 @@
+          .slice(0, -1)
+          .map((segment) => `"${segment.name}"`)
+          .join(', ')
+      : lastSelected;
```
Generally, for these things there is the [Intl.ListFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/ListFormat), but I guess no need to necessarily change now.

```ts
const list = new Intl.ListFormat(language, { style: 'long', type: 'conjunction'})
  .format(allSelected.map((value) => `"${value}"`));
```

JanJakes commented on mailpoet/assets/js/src/segments/dynamic/list/action-confirm.tsx:
```diff
@@ -0,0 +1,194 @@
+      void dispatch(noticesStore).createErrorNotice(errorMessage, {
+        explicitDismiss: true,
+      });
+    });
```
Could we use `async`/`await`?

JanJakes commented on mailpoet/assets/js/src/segments/dynamic/list/get-row.tsx:
```diff
@@ -79,7 +114,13 @@ export function getRow(dynamicSegment: DynamicSegment): object[] {
           </Button>
-          <Icon icon={moreVertical} />
+          <DropdownMenu
+            className="mailpoet-automation-listing-more-button"
```
`mailpoet-automation-listing-...` — "automation" :)

JanJakes commented on mailpoet/assets/js/src/segments/dynamic/list/get-row.tsx:
```diff
@@ -3,11 +3,37 @@ import { DynamicSegment, DynamicSegmentAction } from 'segments/types';
+      error.errors.map(
+        (e) => void dispatch(noticesStore).createErrorNotice(e.message),
+      );
+    });
```
Just noticing the then/catch vs. async/await again.

JanJakes reviewed (commented):
Sorry for possibly a bit messy/outdated comments. The PR contains quite a bit of refactoring of logic written earlier, so some of my comments might be a bit outdated, before I got to the end of the review.

JanJakes commented on mailpoet/assets/js/src/segments/dynamic/list/listing-helpers.ts:
```diff
@@ -25,33 +25,18 @@ export function updateDynamicQuery(values: Partial<DynamicSegmentQuery>): void {
+      query[key] = integerKeys.includes(key) ? parseInt(value, 10) : value;
     }
-  }
+  });
```
I see this was simplified. I think I suggested even a simpler approach [in my comment](https://github.com/mailpoet/mailpoet/pull/5335#discussion_r1497114432), but not sure if it does the same as this one.

Commit:
Remove style causing inconsistent padding

MAILPOET-5395

Commit:
Restore styles that weren't moved over

MAILPOET-5395

Commit:
Make padding consistent with automations listing

MAILPOET-5395

Commit:
Improve readability of plural confirmation messages

MAILPOET-5395

Commit:
Convert segment fetching to await

MAILPOET-5395

Commit:
Convert bulk action to await

MAILPOET-5395

Commit:
Convert duplicate segment method to use await

MAILPOET-5395

johnolek commented:
Thanks as always for your thorough review @JanJakes! Sorry it's taken me some time to get back around to this PR. I've been working on implementing some fixes and I'm hoping to have everything wrapped up tomorrow.

JanJakes committed:
Use PageHeader component

[MAILPOET-5395]

JanJakes committed:
Fix listing header spacing

[MAILPOET-5395]

JanJakes committed:
Fix spacing for segment description

[MAILPOET-5395]

JanJakes committed:
Fix invalid combination of href and disabled props

[MAILPOET-5395]

JanJakes committed:
Fix types for confirmation dialog

[MAILPOET-5395]

JanJakes committed:
Use Intl.ListFormat to format lists

[MAILPOET-5395]

JanJakes committed:
Do not duplicate segments query in store and URL

[MAILPOET-5395]

JanJakes committed:
Use real number of segments for table placeholder

[MAILPOET-5395]

JanJakes committed:
Debounce segment listing search field

[MAILPOET-5395]

JanJakes committed:
Make automation-specific more button style a generic listing style

[MAILPOET-5395]

JanJakes committed:
Make automation-specific listing styles a generic

[MAILPOET-5395]

JanJakes committed:
Add missing border

[MAILPOET-5395]

JanJakes committed:
Remove additional paddings added by paragprahs

[MAILPOET-5395]

JanJakes committed:
Split date & time display in two lines

[MAILPOET-5395]

JanJakes committed:
Improve spacing and display on smaller screens

[MAILPOET-5395]

JanJakes committed:
Unify automation actions styles with segments

[MAILPOET-5395]

JanJakes committed:
Unify automation name styles with segments

[MAILPOET-5395]

JanJakes committed:
Unify segment listing notices with automation listing

[MAILPOET-5395]

JanJakes committed:
Add a way to abort mailpoet ajax requests

[MAILPOET-5395]

JanJakes committed:
Properly handle loading vs empty states

[MAILPOET-5395]

JanJakes committed:
Abort requests when quickly changing listing query

[MAILPOET-5395]

JanJakes committed:
Fix casing of bulk actions label

[MAILPOET-5395]

JanJakes committed:
Remove no longer needed automation tab focus logic

[MAILPOET-5395]

JanJakes committed:
Fix flaky automation test

[MAILPOET-5395]

JanJakes committed:
Fix flaky listing action selection

[MAILPOET-5395]

JanJakes commented:
I added a series of commits starting from https://github.com/mailpoet/mailpoet/pull/5335/commits/09cf50d7d7e5094a2a2a451baa02b09f5864e15b, including:
- reuse existing header component
- unify designs of automation and segment listing
- reuse more styles between automation and segment listing
- use `Intl.ListFormat`
- do not store listing parameters in both URL and Redux store (use URL only)
- debounce search queries
- improve display on smaller screens
- abort listing requests when quickly changing listing query
- fix flaky tests

JanJakes commented:
@websupporter You're actually the original author, so I can't make you a reviewer. So I only assigned it to you. If you're too busy, don't hesitate to reassign.

websupporter commented:
LGTM. Thank you @JanJakes

Aschepikov commented:
Hi @JanJakes @websupporter  I am not sure who will continue with this pr
1. there is an issue with the raws per page, if we choose another number like 50 75 100 there will be an issue that the raws per page still will be 25 but the list of segments will be longer
![image](https://github.com/mailpoet/mailpoet/assets/113185437/5c31e3f5-8906-4be8-b17e-4edab90e8503)
![image](https://github.com/mailpoet/mailpoet/assets/113185437/94b82ce8-e51e-41e2-b26d-5a7ff524383c)
![image](https://github.com/mailpoet/mailpoet/assets/113185437/7c4bb636-b771-4bee-a3a5-32f749b96278)

2. Modified column sorting doesn't work 
![image](https://github.com/mailpoet/mailpoet/assets/113185437/2267d58b-0cc5-4cb6-a985-83b08319f929)
![image](https://github.com/mailpoet/mailpoet/assets/113185437/a4d716e2-891e-47b3-b9ed-40e83c2614e2)

3. There is a 0  with an add a condition button
![image](https://github.com/mailpoet/mailpoet/assets/113185437/9ba58880-c196-48ad-bda3-08b4b228150a)

JanJakes commented:
@Aschepikov I'll check it out and will fix the issues.

JanJakes committed:
Fix "rows per page" not being displayed correcty

[MAILPOET-5395]

JanJakes committed:
Fis "A component is changing an uncontrolled input to be controlled"

[MAILPOET-5395]

JanJakes commented:
@Aschepikov

1. Fixed: https://github.com/mailpoet/mailpoet/pull/5335/commits/f7cee95295cd7d409d3ef1bc65e6f216edb2a1dd
2. Works fine for me, how would I replicate the issue?
3. Good catch, this is unrelated, also on `trunk`, I will create a ticket so we can move on with this.

<img width="239" alt="Screenshot 2024-04-22 at 14 10 39" src="https://github.com/mailpoet/mailpoet/assets/141436/4d40e859-7485-439b-b792-288da91cbd2b">
<img width="217" alt="Screenshot 2024-04-22 at 14 10 43" src="https://github.com/mailpoet/mailpoet/assets/141436/a3486c16-d88e-45cf-b677-6b69034b5248">

Aschepikov reviewed (approved):
GJ! @JanJakes Yep I checked it on a new install and looks like it is connected with the template data
